src/skins/vector/views/organisms/UserSettings.js

```diff
@@ -61,62 +181,115 @@ module.exports = React.createClass({
+                                <div className="mx_UserSettings_profileTableRow">
+                                    <div className="mx_UserSettings_profileLabelCell">
+                                        <label htmlFor="displayName">Display name</label>
+                                    </div>
+                                    <div className="mx_UserSettings_profileInputCell">
+                                        <input id="displayName" ref="displayName" value={ this.state.displayName } onChange={ this.onDisplayNameChange } />
+                                    </div>
                                 </div>
 
-                                <div className="mx_UserSettings_3pids">
-                                    {this.state.threepids.map(function(val) {
-                                        return <div key={val.address}>{val.address}</div>;
-                                    })}
+                                {this.state.threepids.map(function(val) {
+                                    var id = "email-" + val.address;
+                                    return (
+                                        <div className="mx_UserSettings_profileTableRow">
+                                            <div className="mx_UserSettings_profileLabelCell">
+                                                <label htmlFor={ id }>Email</label>
+                                            </div>
+                                            <div className="mx_UserSettings_profileInputCell">
+                                                <input key={val.address} id={ id } value={ val.address } disabled />
+                                            </div>
+                                        </div>
+                                    );
+                                })}
+
+                                <div className="mx_UserSettings_profileTableRow">
+                                    <div className="mx_UserSettings_profileLabelCell">
+                                        <label htmlFor="password1">New password</label>
+                                    </div>
+                                    <div className="mx_UserSettings_profileInputCell">
+                                        <input id="password1" ref="password1" value={ this.state.password1 } />
+                                    </div>
                                 </div>
+                                <div className="mx_UserSettings_profileTableRow">
+                                    <div className="mx_UserSettings_profileLabelCell">
+                                        <label htmlFor="password2">Confirm new password</label>
+                                    </div>
+                                    <div className="mx_UserSettings_profileInputCell">
+                                        <input id="password2" ref="password2" value={ this.state.password2 } />
```

Comment:
It feels like this component is doing way too much. Why should it really have to care about doing confirm fields for passwords *in addition* to all the other stuff the poor thing has to manage? Can we make some of these snippets pure UI components?